actor/src/main/scala/com/evolutiongaming/akkaeffect/util/Serially.scala

```diff
@@ -44,27 +57,33 @@ private[akkaeffect] object Serially {
       def apply(f: A => F[A]) = {
         for {
           d <- Concurrent[F].deferred[Either[Throwable, Unit]]
-          t  = (a: A) => {
+          t = (a: A) => {
             for {
               b <- f(a).attempt
               _ <- d.complete(b.void)
             } yield {
               b.getOrElse(a)
             }
           }
-          a <- ref.modify {
-            case s: S.Idle   =>
-              (S.Active, start(s.value, t))
-            case s: S.Active =>
-              val task = (a: A) => Async[F].defer {
-                for {
-                  a <- s.task(a)
-                  a <- t(a)
-                } yield a
-              }
-              (S.Active(task), Concurrent[F].unit)
-            case S.Active    =>
-              (S.Active(t), Concurrent[F].unit)
+          a <- Sync[F].delay {
+            val s = ref.getAndUpdate {
```

Comment:
`s` is a previous state here, how is it supposed to work?

Reply:
It is only used to select a return value, the same as in original version.

`AtomicReference` does not have `Ref.modify`, so one either have to write spin loop himself, or write "match...case" twice together with `getOrUpdate`.

Reply:
unfortunately one should write spin loop himself…

Reply:
Why the current version is broken? I do not see the bug somehow...

It updates a state to a new one and then uses the old one to select the second part of the tuple. It looks like a full equivalent of the custom spin loop. 🤔

Reply:
I checked once more - I was wrong, it will work!